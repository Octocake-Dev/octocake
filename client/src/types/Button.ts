export type ButtonProps = {
  size?: "sm" | "md" | "lg";
  variant?: "primary" | "danger";
  loading?: boolean;
};
