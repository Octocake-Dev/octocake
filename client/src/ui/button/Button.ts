import { red } from "tailwindcss/colors";
import { variant } from "styled-system";
import styled from "styled-components";

type ButtonProps = {
  size?: "sm" | "md" | "lg";
  variant?: "primary" | "danger";
};

const Button = styled("button")<ButtonProps>(
  {
    cursor: "pointer",
    transition: "250ms ease-in-out",
    fontFamily: "inherit",
    fontWeight: 600,
    borderRadius: "10px",
  },

  () =>
    variant({
      prop: "variant",
      variants: {
        primary: {
          color: "white",
          backgroundColor: "var(--color-primary-500)",

          ":hover": {
            backgroundColor: "var(--color-primary-600)",
          },
          ":focus": {
            backgroundColor: "var(--color-primary-600)",
            boxShadow: `0px 0px 0px 4px rgba(53,143,128,0.4)`,
          },
        },

        danger: {
          color: "white",
          backgroundColor: red[500],

          ":hover": {
            backgroundColor: red[600],
          },
          ":focus": {
            backgroundColor: red[600],
            boxShadow: `0px 0px 0px 4px rgba(220,38,38,0.4)`,
          },
        },
      },
    }),

  variant({
    prop: "size",
    variants: {
      sm: {
        fontSize: "14px",
        padding: "10px 20px",
      },

      md: {
        fontSize: "16px",
        padding: "12px 22px",
      },

      lg: {
        fontSize: "18px",
        padding: "14px 24px",
      },
    },
  })
);

Button.defaultProps = { size: "md", variant: "primary" };

export default Button;
