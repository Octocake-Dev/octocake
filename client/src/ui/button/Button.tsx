import React from "react";

import { ImSpinner2 } from "react-icons/im";

import StyledButton from "./Button.styles";

import type { ButtonProps } from "@/types/Button";

const Button = ({ children, disabled, loading, ...rest }: ButtonProps) => (
  <StyledButton disabled={disabled || loading} {...rest}>
    {children}

    {loading && (
      <span className="ml-2">
        <ImSpinner2 className="animate-spin" />
      </span>
    )}
  </StyledButton>
);

Button.defaultProps = { size: "md", variant: "primary" };

export default Button;
