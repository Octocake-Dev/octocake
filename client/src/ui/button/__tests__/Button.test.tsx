/**
 * @jest-environment jsdom
 */
import React from "react";

import { render } from "@testing-library/react";

import Button from "@/ui/button/Button";

describe("Button", () => {
  it("should render a button with Click me text", () => {
    const { getByText } = render(<Button>Click me</Button>);

    const text = getByText("Click me");

    expect(text).toBeVisible();
  });
});
