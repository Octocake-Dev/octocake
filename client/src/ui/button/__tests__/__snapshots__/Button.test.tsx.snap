// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Should match snapshot 1`] = `
<button
  class="Button-sc-1chh247-0 eOzvOL"
  disabled=""
>
  Click me
</button>
`;
